examples/nlp/mlm_with_bert.py

```diff
@@ -0,0 +1,209 @@
+    # set targets to -1 by default, it means ignore
+    labels = -1 * np.ones(X.shape, dtype=int)
+    # set labels for masked tokens
+    labels[inp_mask] = X[inp_mask]
+
+    # prepare input
+    X_mlm = np.copy(X)
+    # set input to [MASK] which is the last token for the 90% of tokens
+    # this means leaving 10% unchanged
+    inp_mask_2mask = inp_mask & (np.random.rand(*X.shape) < 0.90)
+    X_mlm[
+        inp_mask_2mask
+    ] = tokenizer.mask_token_id  # mask token is the last in the dict
+
+    # set 10% to a random token
+    inp_mask_2random = inp_mask_2mask & (np.random.rand(*X.shape) < 1 / 9)
+    X_mlm[inp_mask_2random] = np.random.randint(
+        3, tokenizer.mask_token_id, inp_mask_2random.sum()
+    )
+
+    return X_mlm, labels
+
+
+# use validation and test data for mlm
+X_train_mlm = np.vstack(X_data)
+# masks and labels
+X_train_mlm, y_train_mlm = prepare_mlm_input_and_labels(X_train_mlm)
+
+
+"""
+## Create MaskedLanguageModel using huggingface transformers
+"""
+
+
+def masked_sparse_categorical_crossentropy(y_true, y_pred):
+    y_true_masked = tf.boolean_mask(y_true, tf.not_equal(y_true, -1))
+    y_pred_masked = tf.boolean_mask(y_pred, tf.not_equal(y_true, -1))
+    loss = tf.keras.losses.sparse_categorical_crossentropy(
+        y_true_masked, y_pred_masked, from_logits=True
+    )
+    return loss
+
+
+class MaskedLanguageModel(tf.keras.Model):
+    def train_step(self, inputs):
+        features, labels = inputs
+
+        with tf.GradientTape() as tape:
+
+            predictions = self(features, training=True)[0]
+            loss = masked_sparse_categorical_crossentropy(labels, predictions)
+
+        # Compute gradients
+        trainable_vars = self.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+
+        # Update weights
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Compute our own metrics
+        loss_tracker.update_state(loss)
+
+        # Return a dict mapping metric names to current value
+        return {"loss": loss_tracker.result()}
+
+
+with strategy.scope():
+    loss_tracker = tf.keras.metrics.Mean(name="loss")
+    input_layer = tf.keras.layers.Input((flags.MAX_LEN,), dtype=tf.int32)
+    bert_model = TFAutoModelWithLMHead.from_pretrained(flags.PRETRAINED_MODEL)
+    output_layer = bert_model(input_layer)
+    mlm_model = MaskedLanguageModel(input_layer, output_layer)
+
+    optimizer = tf.keras.optimizers.Adam(learning_rate=flags.LR)
+    mlm_model.compile(optimizer=optimizer)
+
+mlm_model.summary()
+
+"""
+## Train and Save
+"""
+
+mlm_model.fit(X_train_mlm, y_train_mlm, epochs=3, batch_size=global_batch_size)
+
+# Save trained model using transfomers .save_pretrained()
+bert_model.save_pretrained("imdb_bert_uncased")
+
+"""
+## Load and Test
+"""
+
+imdb_bert_model = TFAutoModelWithLMHead.from_pretrained("imdb_bert_uncased")
+nlp = pipeline("fill-mask", model=imdb_bert_model, tokenizer=tokenizer, framework="tf")
```

Comment:
A concern I have here: this example is basically more about "how to use HuggingFace Transformers" than about masked language modeling or Keras. The pipeline, tokenizer, and `TFAutoModelWithLMHead` are black boxes.

The purpose of our code examples is to be Keras-centric and as didactic as possible, i.e. they should explain the task and problem domain, similar to [what we do here for object detection](https://keras.io/examples/vision/retinanet/), where everything is implemented from scratch.

Would you consider implementing these component from scratch in Keras, e.g. with the `TextVectorization` layer for tokenization and with a Transformer block built with the new `MultiHeadAttention` layer (in tf-nightly)? No pretraining necessary.

The model could be similar to [this](https://keras.io/examples/generative/text_generation_with_miniature_gpt/), expect you wouldn't even have to implement `MultiHeadAttention`.

Reply:
Hi, @fchollet Thanks for your comment. 

True, I feel the same, this example looks like 'how to use HuggingFace Transformers'. 

Yes, I would like to consider implementing these components from scratch in Keras, e.g. with the TextVectorization layer for tokenization and with a Transformer block built with the new MultiHeadAttention layer. 

Should I use this for the classification Problems or Pretraining Problems?

Reply:
I have started according to your suggestion to use MultiHead layer and TextVectorizationLayer. And it's almost done for both pretraning and fine tuning. 

I have one doubt, TextVectorizationLayer is not running on TPU. Same notebook running on CPU n GPU, but not on TPU

Reply:
Are you using `TextVectorization` as part of the tf.data pipeline? Can you share your notebook?

Reply:
No, I am not using tf.data. 

```
def get_vectorize_layer(texts, vocab_size, max_seq, special_tokens=['[MASK]']):

  vectorize_layer = preprocessing.TextVectorization(
    max_tokens=vocab_size,
    output_mode='int',
    standardize = None,
    output_sequence_length=max_seq)
  
  vectorize_layer.adapt(texts) # error in adapt

  vocab = vectorize_layer.get_vocabulary()
  vocab = vocab[2:vocab_size-len(special_tokens)] + ['[MASK]']
  vectorize_layer.set_vocabulary(vocab)
  return vectorize_layer


vectorize_layer = get_vectorize_layer(data.text.values.tolist(), 20000, 196, special_tokens=['[MASK]'])
```

Error is:

> NotFoundError: 'OptimizeDatasetV2' is neither a type of a primitive operation nor a name of a function registered in binary running on n-86c78cbc-w-0. Make sure the operation or function is registered in the binary running in this process.

Tensorflow: 2.4.0-dev20200915
Reference Notebook: https://colab.research.google.com/drive/1QiQmA2M2WkwsjiHHxBhPM0Jw3_LBJnGV?usp=sharing

**Note: Same code working with tensorflow:2.3.0**

Reply:
This is a bug likely to a recently introduced optimization setting for tf.data datasets (which are used internally in `adapt`). We're looking into it -- thank you for the report.

Since we will fix this before the next release, I recommend that you keep developing this example with `TextVectorization`, using GPU only for now. You could try to use an older tf-nightly version that wouldn't include the bug (though I am not sure if you'll find one that also has the MultiHeadAttention layer).

Reply:
I am developing with GPU compatible. I have tried almost random 5-7 different tf-nightly version. All have same issues. 

My notebook is almost ready. going to commit in a few mins.
Thanks!